### PR TITLE
Not for submission: First pass at SessionPool (and PooledSession)

### PR DESCRIPTION
**Pull request by jskeet**
This is unused as yet.

This PR has no unit tests yet, and is to allow for early review of the API and any immediate problems, while I write unit tests.

### CODE REVIEW

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        public Task<Transaction> BeginTransactionAsync(BeginTransactionRequest request, int timeoutSeconds, CancellationToken cancellationToken)
+        {
+            CheckNotDisposed();
+            GaxPreconditions.CheckNotNull(request, nameof(request));
+            request.SessionAsSessionName = SessionName;
+            var settings = CreateSettings(Client.Settings.BeginTransactionSettings, timeoutSeconds, cancellationToken);
+            return RecordSuccessAndExpiredSessions(Client.BeginTransactionAsync(request, settings));
+        }
+
+        private async Task<T> RecordSuccessAndExpiredSessions<T>(Task<T> task, [CallerMemberName] string caller = null)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs]: I've simplified this method to remove all the logging bits - they were when I was trying to diagnose something else.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            }
+            var settings = CreateSettings(Client.Settings.ExecuteSqlSettings, timeoutSeconds, cancellationToken);
+            return RecordSuccessAndExpiredSessions(Client.ExecuteSqlAsync(request, settings));
+        }
+
+        /// <summary>
+        /// Executes a BeginTransaction RPC asynchronously.
+        /// </summary>
+        /// <remarks>
+        /// This method does not affect <see cref="TransactionId"/> of this object.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs]: Is this like a nested transaction then?

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs]: OK, but you haven't pushed the commit with the simplification yet, right?
Or else maybe remove the `StringBuilder log` as well?

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        /// </summary>
+        private PooledSession AfterReset() => new PooledSession(_pool, SessionName, null, ModeOneofCase.None, _evictionTime, RefreshTicks);
+
+        internal PooledSession WithTransaction(ByteString transactionId, ModeOneofCase transactionMode) =>
+            new PooledSession(_pool, SessionName, transactionId, transactionMode, _evictionTime, _refreshTicks);
+
+        /// <summary>
+        /// Indicates whether the associated session has expired and should be evicted from the pool.
+        /// </summary>
+        internal bool ShouldBeEvicted => _pool.Options.Clock.GetCurrentDateTimeUtc() > _evictionTime;
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs]: OK, removing my comment on MaintainPool as well.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs]: I like the separation of RPC and management.

But this method is `public` and `WithTransaction` is `internal`, I think they both need to be `public` and this one needs more documentation, or they both need to be `internal` and only be called from `SessionPool` (I prefer both `internal`).
Right now nothing, not even docs, is preventing client code from calling `BeginTransactionAsync` and then `Commit` and that, if the original transaction was stomped, will definetely failed.

Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PoolRewrite/PooledSession.cs]: Either way works better than how it is now, I prefer `internal` but I'm really fine with either, really.

**Comment by jskeet** [comment]
Closing *this* PR for now, as we're going to split this up further. Not just updating it with new commits, as we want all the conversation.

**Pull request closed without merging**